### PR TITLE
Image manager and return NSData for GIFs

### PR DESCRIPTION
**Pull request by nihalahmed**

### CODE REVIEW

**Comment by gyfelton** [comment]
Remember to update changelog,d :)

File: SDWebImage/SDImageCache.h [hunk 1]
```diff
@@ -63,6 +66,11 @@ typedef void(^SDWebImageCalculateSizeBlock)(NSUInteger fileCount, NSUInteger tot
 @property (assign, nonatomic) NSUInteger maxCacheSize;
 
 /**
+ * If set to YES, last access date will be used when cleaning disk.
```
Comment (gyfelton) [inline SDWebImage/SDImageCache.h]: if no, then use last modified date?

File: SDWebImage/SDImageCache.h [hunk 2]
```diff
@@ -125,6 +133,19 @@ typedef void(^SDWebImageCalculateSizeBlock)(NSUInteger fileCount, NSUInteger tot
 - (void)storeImage:(UIImage *)image recalculateFromImage:(BOOL)recalculate imageData:(NSData *)imageData forKey:(NSString *)key toDisk:(BOOL)toDisk;
 
 /**
+ * Store an image into memory and optionally disk cache at the given key.
+ *
+ * @param image       The image to store
```
Comment (gyfelton) [inline SDWebImage/SDImageCache.h:138]: either in UIImage or NSData representation

File: SDWebImage/SDImageCache.h [hunk 3]
```diff
@@ -132,6 +153,14 @@ typedef void(^SDWebImageCalculateSizeBlock)(NSUInteger fileCount, NSUInteger tot
 - (NSOperation *)queryDiskCacheForKey:(NSString *)key done:(SDWebImageQueryCompletedBlock)doneBlock;
 
 /**
+ * Query the disk cache asynchronously.
+ *
+ * @param key               The unique key used to store the wanted image
+ * @param returnDataForGIFs Return NSData for GIFs if YES
```
Comment (gyfelton) [inline SDWebImage/SDImageCache.h]: mismatch param name, don't understand what to pass in here, a SDWebImageReturnConfig instance right?

File: SDWebImage/SDImageCache.h [hunk 4]
```diff
@@ -139,6 +168,14 @@ typedef void(^SDWebImageCalculateSizeBlock)(NSUInteger fileCount, NSUInteger tot
 - (UIImage *)imageFromMemoryCacheForKey:(NSString *)key;
 
 /**
+ * Query the memory cache synchronously.
+ *
+ * @param key               The unique key used to store the wanted image
+ * @param returnDataForGIFs Return NSData for GIFs if YES
```
Comment (gyfelton) [inline SDWebImage/SDImageCache.h]: same comment here

File: SDWebImage/SDImageCache.m [hunk 5]
```diff
@@ -312,15 +337,18 @@ - (NSData *)diskImageDataBySearchingAllPathsForKey:(NSString *)key {
     return nil;
 }
 
-- (UIImage *)diskImageForKey:(NSString *)key {
+- (SDImage *)diskImageForKey:(NSString *)key returnConfig:(SDWebImageReturnConfig *)returnConfig {
     NSData *data = [self diskImageDataBySearchingAllPathsForKey:key];
     if (data) {
+        if (returnConfig.returnDataForGIFs && [NSData sd_isContentTypeGIFForImageData:data]) {
```
Comment (gyfelton) [inline SDWebImage/SDImageCache.m]: break into two lines

File: SDWebImage/SDWebImageDownloader.h [hunk 6]
```diff
@@ -179,6 +182,33 @@ typedef NSDictionary *(^SDWebImageDownloaderHeadersFilterBlock)(NSURL *url, NSDi
...
 
 /**
+ * Creates a SDWebImageDownloader async downloader instance with a given URL
+ *
+ * The delegate will be informed when the image is finish downloaded or an error has happen.
+ *
+ * @see SDWebImageDownloaderDelegate
+ *
+ * @param url            The URL to the image to download
+ * @param options        The options to be used for this download
```
Comment (gyfelton) [inline SDWebImage/SDWebImageDownloader.h:192]: missing the returnConfig param

File: SDWebImage/SDWebImageDownloader.m [hunk 7]
```diff
@@ -191,7 +198,8 @@ - (void)setOperationClass:(Class)operationClass {
     return operation;
 }
 
-- (void)addProgressCallback:(SDWebImageDownloaderProgressBlock)progressBlock andCompletedBlock:(SDWebImageDownloaderCompletedBlock)completedBlock forURL:(NSURL *)url createCallback:(SDWebImageNoParamsBlock)createCallback {
+- (void)addProgressCallback:(SDWebImageDownloaderProgressBlock)progressBlock andCompletedBlock:(SDWebImageDownloaderCompletedBlock)completedBlock forURL:(NSURL *)url createCallback:(SDWebImageNoParamsBlock)createCallback
+{
```
Comment (gyfelton) [inline SDWebImage/SDWebImageDownloader.m]: revert this change

File: SDWebImage/SDWebImageDownloaderOperation.m [hunk 8]
```diff
@@ -381,6 +388,11 @@ - (void)connectionDidFinishLoading:(NSURLConnection *)aConnection {
         if (self.options & SDWebImageDownloaderIgnoreCachedResponse && responseFromCached) {
             completionBlock(nil, nil, nil, YES);
         } else if (self.imageData) {
+            if (self.returnConfig.returnDataForGIFs && [NSData sd_isContentTypeGIFForImageData:self.imageData]) {
```
Comment (gyfelton) [inline SDWebImage/SDWebImageDownloaderOperation.m]: break into two lines

**Comment by gyfelton** [comment]
rc

File: SDWebImage/SDImageCache.h (hunk 1 above)
Comment (nihalahmed) [inline SDWebImage/SDImageCache.h]: yes

**Comment by nihalahmed** [comment]
@gyfelton another look?

File: SDWebImage/SDImageCache.h (hunk 1 above)
Comment (gyfelton) [inline SDWebImage/SDImageCache.h]: better put the comment here to clarify thing

**Comment by gyfelton** [comment]
:shipit:
